core/client/batch_client.go

```diff
@@ -86,34 +88,71 @@ func (c *Client) BatchCreateMutation(ctx context.Context, users []*User) ([]*ent
 	if err != nil {
 		return nil, err
 	}
-	mutations := make([]*entry.Mutation, 0, len(users))
 
-	for _, u := range users {
-		leaf, ok := leavesByUserID[u.UserID]
-		if !ok {
-			return nil, fmt.Errorf("no leaf found for %v", u.UserID)
+	type result struct {
+		m   *entry.Mutation
+		err error
+	}
+	uChan := make(chan *User)
```

Comment:
I don't see any particular reason a buffer would help.  There should be N readers executing at any time, and we're not trying to avoid blocking of the allocation function which is in it's own thread.

Reply:
You'll generally get better performance because you increase the potential for concurrency:

code: https://play.golang.org/p/b-yXwNX2VDx
```
BenchmarkNoBuffer-6           	 2000000	       809 ns/op
BenchmarkWithBufferOf10-6     	 3000000	       518 ns/op
BenchmarkWithBufferof100-6    	 5000000	       286 ns/op
BenchmarkWithBufferof1000-6   	 5000000	       297 ns/op
```

Might not be super important in this context, but it's worth being aware of, and I think there should be a reason _not_ to buffer, rather than the other way around.